PowerShell extension doesn't start on slow machines

### Prerequisites

- [X] I have written a descriptive issue title.
- [X] I have searched all [_open and closed_ issues](https://github.com/PowerShell/vscode-powershell/issues?q=is%3Aissue) to ensure it has not already been reported.
- [X] I have read the [troubleshooting](https://github.com/PowerShell/vscode-powershell/blob/main/docs/troubleshooting.md) guide.
- [X] I am sure this issue is with the _extension itself_ and does not reproduce in a standalone [PowerShell](https://github.com/PowerShell/PowerShell/issues/new/choose) instance.
- [X] I have verified that I am using the latest version of Visual Studio Code and the PowerShell extension.
- [X] If this is a security issue, I have read the [security issue reporting guidance](https://github.com/PowerShell/vscode-powershell/blob/main/SECURITY.md).

### Summary

Whenever I start VS Code for the first time in a Windows session, PowerShell extension starts with a warning.

Moreover, closing the terminal session results in the below mentioned error.

Plus, the PowerShell extension banner message is looking odd:

![PowerShell extension banner](https://github.com/user-attachments/assets/c2002d3d-f764-4690-9fa9-ded2557c877a)

### PowerShell Version

```console
Name                           Value
----                           -----
PSVersion                      7.4.4
PSEdition                      Core
GitCommitId                    7.4.4
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0.}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```


### Visual Studio Code Version

```console
1.92.1
x64
```


### Extension Version

```console
ms-vscode.powershell@2024.2.2
```


### Steps to Reproduce

1. Launch a Windows session on a slow machine
2. Right-click a PowerShell file and select "Open in Visual Studio Code" context menu item.
3. Immediately open the terminal window using the keyboard shortcut <CTRL>+<Ö>.

### Visuals

_No response_

### Logs

```
21.8.2024 16:17:03 [NORMAL] - Visual Studio Code: v1.92.1 64-bit on Windows 64-bit

21.8.2024 16:17:03 [NORMAL] - PowerShell Extension: v2024.2.2

21.8.2024 16:17:03 [NORMAL] - Starting 'PowerShell (x64)' at: C:\Program Files\PowerShell\7\pwsh.exe

21.8.2024 16:17:04 [NORMAL] - PowerShell process started with PID: 9012

21.8.2024 16:17:07 [WARNING] - PowerShell process terminated or Extension Terminal was closed, PID: 9012

21.8.2024 16:17:08 [ERROR] - Extension Terminal is undefined.

21.8.2024 16:17:09 [ERROR] - PowerShell Language Server process didn't start!
```

--- a/src/process.ts
+++ b/src/process.ts
@@ -22,6 +22,10 @@ export class PowerShellProcess {
     private pid?: number;
     private pidUpdateEmitter?: vscode.EventEmitter<number | undefined>;
 
+    // Captured when the terminal closes so callers can report why the process
+    // exited (e.g. a non-zero exit code) instead of a generic failure message.
+    private exitStatus?: vscode.TerminalExitStatus;
+
     constructor(
         public exePath: string,
         private bundledModulesPath: string,
@@ -223,6 +227,13 @@ export class PowerShellProcess {
         return await this.consoleTerminal?.processId;
     }
 
+    // Returns the exit status of the terminal, if it has closed. This is
+    // captured before the terminal is disposed so callers can report a non-zero
+    // exit code when the process terminates before connecting.
+    public getExitStatus(): vscode.TerminalExitStatus | undefined {
+        return this.exitStatus;
+    }
+
     public showTerminal(preserveFocus?: boolean): void {
         this.consoleTerminal?.show(preserveFocus);
     }
@@ -343,8 +354,16 @@ export class PowerShellProcess {
             return;
         }
 
+        // Capture the exit status before disposing so it can be reported. A
+        // non-zero code means the process failed to start; `undefined` means
+        // the user (or VS Code) closed the terminal.
+        this.exitStatus = terminal.exitStatus;
+        const code = this.exitStatus?.code;
+
         this.logger.writeWarning(
-            `PowerShell process terminated or Extension Terminal was closed, PID: ${this.pid}`,
+            `PowerShell process terminated or Extension Terminal was closed${
+                code !== undefined ? ` with exit code: ${code}` : ""
+            }, PID: ${this.pid}`,
         );
         this.dispose();
     }

--- a/src/session.ts
+++ b/src/session.ts
@@ -791,8 +791,15 @@ export class SessionManager implements Middleware {
             );
         } else {
             shouldUpdate = false;
+            // If the process terminated before connecting, VS Code gives us the
+            // terminal's exit code, which is far more actionable than a generic
+            // failure. A `undefined` code means it timed out or was closed
+            // without reporting one.
+            const exitCode = powerShellProcess.getExitStatus()?.code;
             void this.setSessionFailedOpenBug(
-                "PowerShell Language Server process didn't start!",
+                exitCode !== undefined && exitCode !== 0
+                    ? `PowerShell Language Server process exited with code: ${exitCode} before connecting! Check the PowerShell output for errors.`
+                    : "PowerShell Language Server process didn't start!",
             );
         }
 

--- a/test/core/process.test.ts
+++ b/test/core/process.test.ts
@@ -1,0 +1,90 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import * as assert from "assert";
+import Sinon from "sinon";
+import * as vscode from "vscode";
+import { PowerShellProcess } from "../../src/process";
+import { stubInterface, testLogger } from "../utils";
+
+describe("PowerShellProcess", () => {
+    afterEach(() => {
+        Sinon.restore();
+    });
+
+    function createProcess(): PowerShellProcess {
+        return new PowerShellProcess(
+            "pwsh",
+            "modules",
+            false,
+            false,
+            testLogger,
+            vscode.Uri.file("C:/tmp"),
+            "",
+            vscode.Uri.file("C:/tmp/session.json"),
+        );
+    }
+
+    // These pokes mirror the `manager as unknown as {...}` style used in
+    // `session.test.ts` to exercise otherwise private startup behavior.
+    function internalsOf(process: PowerShellProcess): {
+        consoleTerminal: vscode.Terminal | undefined;
+        onTerminalClose: (terminal: vscode.Terminal) => void;
+    } {
+        return process as unknown as {
+            consoleTerminal: vscode.Terminal | undefined;
+            onTerminalClose: (terminal: vscode.Terminal) => void;
+        };
+    }
+
+    it("has no exit status before its terminal closes", () => {
+        const process = createProcess();
+        assert.equal(process.getExitStatus(), undefined);
+    });
+
+    it("captures and logs the exit status when its terminal closes", () => {
+        const warnSpy = Sinon.spy(testLogger, "writeWarning");
+        const process = createProcess();
+        const exitStatus: vscode.TerminalExitStatus = {
+            code: 1,
+            reason: vscode.TerminalExitReason.Process,
+        };
+        const terminal = stubInterface<vscode.Terminal>({
+            exitStatus,
+            dispose: () => {
+                return;
+            },
+        });
+
+        // Track the terminal, then simulate VS Code firing its close event.
+        const internals = internalsOf(process);
+        internals.consoleTerminal = terminal;
+        internals.onTerminalClose(terminal);
+
+        assert.deepEqual(process.getExitStatus(), exitStatus);
+        assert.ok(
+            warnSpy.calledWithMatch("exit code: 1"),
+            "should log the non-zero exit code",
+        );
+    });
+
+    it("ignores close events from unrelated terminals", () => {
+        const process = createProcess();
+        const internals = internalsOf(process);
+        internals.consoleTerminal = stubInterface<vscode.Terminal>({
+            dispose: () => {
+                return;
+            },
+        });
+
+        const otherTerminal = stubInterface<vscode.Terminal>({
+            exitStatus: {
+                code: 1,
+                reason: vscode.TerminalExitReason.Process,
+            },
+        });
+        internals.onTerminalClose(otherTerminal);
+
+        assert.equal(process.getExitStatus(), undefined);
+    });
+});

